src/main/java/org/entando/plugins/pda/pam/service/task/KieTaskService.java

```diff
@@ -2,79 +2,166 @@
 import org.entando.web.response.PagedMetadata;
 import org.entando.web.response.PagedRestResponse;
 import org.springframework.boot.web.client.RestTemplateBuilder;
 import org.springframework.stereotype.Service;
+import org.springframework.web.client.HttpServerErrorException;
 import org.springframework.web.client.RestTemplate;
 
 @Service
+@Slf4j
 @RequiredArgsConstructor
 public class KieTaskService implements TaskService {
 
     @NonNull
     private RestTemplateBuilder restTemplateBuilder;
 
     //CHECKSTYLE:OFF
-    public static final String TASK_LIST_URL = "/services/rest/server/queries/tasks/instances/pot-owners";
-    public static final String TASK_PROPERTIES_URL = "/services/rest/server/queries/processes/instances/{pInstanceId}/variables/instances";
+    public static final String TASK_LIST_URL = "/server/queries/tasks/instances/pot-owners";
+    public static final String TASK_URL = "/server/queries/tasks/instances/{tInstanceId}";
+    public static final String TASK_DETAILS_URL = "/server/containers/{containerId}/tasks/{tInstanceId}?withInputData=true&withOutputData=true";
+    public static final String PROCESS_VARIABLES_URL = "/server/queries/processes/instances/{pInstanceId}/variables/instances";
     //CHECKSTYLE:ON
 
+    @Override
     public PagedRestResponse<Task> list(Connection connection, PagedListRequest request) {
         RestTemplate restTemplate = restTemplateBuilder
-                .basicAuthentication(connection.getUsername(), connection.getPassword())
+                .basicAuthorization(connection.getUsername(), connection.getPassword())
                 .build();
 
-        KieTasksResponse response = Optional.ofNullable(restTemplate.getForObject(
-                connection.getUrl() + TASK_LIST_URL, KieTasksResponse.class))
-                .orElseThrow(BadRequestException::new);
-
-        List<Task> result = Optional.ofNullable(response.getTasks())
-                .orElse(Collections.emptyList())
-                .stream()
-                .map(this::fromDto)
-                .peek((Task t) -> {
-                    KieProcessVariablesResponse variablesResponse = Optional.ofNullable(restTemplate.getForObject(
-                            connection.getUrl() + TASK_PROPERTIES_URL,
-                            KieProcessVariablesResponse.class, t.getProcessInstanceId()))
-                            .orElseThrow(BadRequestException::new);
-
-                    t.setProperties(Optional.ofNullable(variablesResponse.getVariables())
-                                        .orElse(Collections.emptyList())
-                            .stream().map(v -> new AbstractMap.SimpleEntry<>(v.getName(), v.getValue()))
+        final Map<Integer, List<KieProcessVariable>> cachedVariables = new ConcurrentHashMap<>();
+
+        List<Task> result = getTasks(restTemplate, connection, request).stream() //Get Tasks
+                .parallel()
+                .peek(t -> { //Get Task Details
```

Comment:
According to Javadoc for peek method: "This method exists mainly to support debugging, where you want to see the elements as they flow past a certain point in a pipeline". Also, in this article https://4comprehension.com/idiomatic-peeking/ It says: "It seems to work fine when applying mutating operations but should not be used this way because this behaviour is non-deterministic due to the possibility of certain peek() calls being omitted due to internal optimization". Could you verify if there is another way to implement this functionality without peek method?

Reply:
I was using `map`, and IntelliJ alerted me that if the item is not being changed I should use `peek`.